### DISCUSSION TITLE
pinocchio3-preview undefined reference when compiling

### DISCUSSION THREAD

**Discussion by giulioturrisi** [stack-of-tasks/pinocchio#2018] [Q&A] [answered]
Hi all, 

I was trying to compile pinocchio3-preview from source. I'm using a docker image with ubuntu 22.04 (where I use a recently forked repo of Pinocchio to modify only the eigenpy version - from 2.7.0 to 2.9.2). I installed casadi 3.6.3 and libbost 1.74.0.3ubuntu7, and activated the casadi support during cmake.

During compilation, around 85%, I get the following error:

```
/usr/bin/ld: CMakeFiles/test-cpp-casadi-basic.dir/basic.cpp.o: in function `boost::unit_test::make_test_case(boost::function<void ()> const&, boost::unit_test::basic_cstring<char const>, boost::unit_test::basic_cstring<char const>, unsigned long) [clone .isra.0]':
basic.cpp:(.text.startup+0x41): undefined reference to `boost::unit_test::ut_detail::normalize_test_case_name(boost::unit_test::basic_cstring<char const>)'
collect2: error: ld returned 1 exit status
```

Do you know if there is some incompatibility between the version of boost and casadi that I'm using? Or if this problem could be related to something else.

### Docker image

The docker image file can be found here: 
```
https://github.com/giulioturrisi/Flywheel-Pendulum-Robot/blob/main/docker_files/integrated_gpu/dockerfile
```

To reproduce the error, you just need to build it, and automatically Pinocchio will be compiled.

### System

- Ubuntu 22.04
- Pinocchio version: pinocchio3-preview

Thank you in advance!

**Comment by nim65s** [comment, marked as answer]
Hi,

Thanks for the report.

At first glance, I'd say this is because:
1. you used `pip install casadi` 
2. this resolve to  [`casadi-3.6.3-cp310-none-manylinux2014_x86_64.whl` ](https://files.pythonhosted.org/packages/76/22/b3651c315af098300ed3b0d43ad7d2ffb300f9aaa6668df11a3f8b484be6/casadi-3.6.3-cp310-none-manylinux2014_x86_64.whl)
3. which was compiled on the `manylinux2014` distribution 
4. which is derivated from CentOS 7
5. which by default compile packages for the old c++ ABI
6. while you are on a ubuntu jammy
7. where packages are by default compiled for the new c++11 ABI.

Therefore, `std::string` signature is not the same between your different binaries.

Possible workarounds (choose one):
1. install robotpkg-casadi, as you already have robotpkg, and robotpkg binaries are compiled with the correct compiler for each distribution
2. compile casadi yourself to get everything with `-D_GLIBCXX_USE_CXX11_ABI=1` (the default on your new system)
3. keep casadi, and compile everything else using `-D_GLIBCXX_USE_CXX11_ABI=0` (this mean you can't have robotpkg binaries anymore)
4. directly ask for a working dockerfile and provide the list of package you want
5. switch to conda (but I won't be helpful on this path for now, and I'm not even sure it has all the packages you need)

**Reply by nim65s** [reply]
Long term solution: improve casadi wheels. I already spent a huge week of this, this is way from done.

**Reply by giulioturrisi** [reply]
Problem solved following your suggestion.

Thank you!
